freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,210 @@
+"""
+Fetch daily-archived OHLCV data from https://data.binance.vision/
+"""
+
+import asyncio
+import datetime
+import io
+import logging
+import zipfile
+
+import aiohttp
+import pandas as pd
+from pandas import DataFrame
+
+from freqtrade.enums import CandleType
+from freqtrade.misc import chunks
+from freqtrade.util.datetime_helpers import dt_from_ts, dt_now
+
+
+logger = logging.getLogger(__name__)
+
+
+class BadHttpStatus(Exception):
+    """Not 200/404"""
+
+    pass
+
+
+async def fetch_ohlcv(
+    candle_type: CandleType,
+    pair: str,
+    timeframe: str,
+    since_ms: int,
+    until_ms: int | None,
+    stop_on_404: bool = False,
+) -> DataFrame:
+    """
+    Fetch OHLCV data from https://data.binance.vision/
+    :candle_type: Currently only spot and futures are supported
+    :param until_ms: `None` indicates the timestamp of the latest available data
+    :param stop_on_404: Stop to download the following data when a 404 returned
+    :return: the date range is between [since_ms, until_ms),
+        return None if no data available in the time range
+    """
+    if candle_type == CandleType.SPOT:
+        asset_type = "spot"
+    elif candle_type == CandleType.FUTURES:
+        asset_type = "futures/um"
+    else:
+        raise ValueError(f"Unsupported CandleType: {candle_type}")
+    symbol = symbol_ccxt_to_binance(pair)
+    start = dt_from_ts(since_ms)
+    end = dt_from_ts(until_ms) if until_ms else dt_now()
+
+    # We use two days ago as the last available day because the daily archives are daily uploaded
+    # and have several hours delay
+    last_available_date = dt_now() - datetime.timedelta(days=2)
+    end = min(end, last_available_date)
+    if start >= end:
+        return DataFrame()
+    df = await _fetch_ohlcv(asset_type, symbol, timeframe, start, end, stop_on_404)
+    logger.info(
+        f"Downloaded data for {pair} from https://data.binance.vision/ with length {len(df)}."
+    )
+    if not df.empty:
+        return df.loc[(df["date"] >= start) & (df["date"] < end)]
+    else:
+        return df
+
+
+def symbol_ccxt_to_binance(symbol: str) -> str:
+    """
+    Convert ccxt symbol notation to binance notation
+    e.g. BTC/USDT -> BTCUSDT, BTC/USDT:USDT -> BTCUSDT
+    """
+    if ":" in symbol:
+        parts = symbol.split(":")
+        if len(parts) != 2:
+            raise ValueError(f"Cannot recognize symbol: {symbol}")
+        return parts[0].replace("/", "")
+    else:
+        return symbol.replace("/", "")
+
+
+def concat(dfs) -> DataFrame:
+    if all(df is None for df in dfs):
+        return DataFrame()
+    else:
+        return pd.concat(dfs)
+
+
+async def _fetch_ohlcv(
+    asset_type: str,
```

Comment:
You might want to improve the naming of this variable
It suggests "spot" / "futures" - but actually contains the directory we intend to download ... which is `futures/um` ... so it's more like "asset_type_url_part - or something similar ?

Reply:
Naming is always the hardest part ......

Reply:
Renamed to `asset_type_url_segment`